core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -516,61 +517,131 @@ class KafkaApis(val requestChannel: RequestChannel,
   def handleOffsetRequest(request: RequestChannel.Request) {
     val correlationId = request.header.correlationId
     val clientId = request.header.clientId
+    val version = request.header.apiVersion()
     val offsetRequest = request.body.asInstanceOf[ListOffsetRequest]
 
-    val (authorizedRequestInfo, unauthorizedRequestInfo) = offsetRequest.offsetData.asScala.partition {
-      case (topicPartition, _) => authorize(request.session, Describe, new Resource(auth.Topic, topicPartition.topic))
-    }
+    val mergedResponseMap = {
```

Comment:
nit: this block is pretty huge. Since each version is handled independently. Perhaps we can move the respective handling logic into separate functions?